cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -189,6 +191,72 @@ Status ConstBoolExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
   return Status::OK();
 }
 
+struct NonZeroVisitor {
+   UInt32Builder *builder;
```

Comment:
AFAIK, indices in Arrow are usually represented with `int64_t`. Also, it might be the case that a different integer type is desired, so can these be generalized? Maybe using `NumericBuilder<...>` but these would not allow using `BooleanBuilder`.